spyder_kernels/customize/spydercustomize.py

```diff
@@ -303,16 +285,34 @@ def _patched_preparation_data(name):
 #==============================================================================
 # Pdb adjustments
 #==============================================================================
-class SpyderPdb(pdb.Pdb):
+class SpyderPdb(pdb.Pdb, object):  # Inherits `object` to call super() in PY2
 
     send_initial_notification = True
     starting = True
+    def __init__(self, *args, **kwargs):
+        """Init Pdb."""
+        self.continue_if_has_breakpoints = True
+        super(SpyderPdb, self).__init__(*args, **kwargs)
 
     # --- Methods overriden by us
     def preloop(self):
         """Ask Spyder for breakpoints before the first prompt is created."""
-        if self.starting:
-            get_ipython().kernel._ask_spyder_for_breakpoints()
+        try:
+            _frontend_request(blocking=False).set_debug(True)
+            if self.starting:
+                breakpoints = _frontend_request().get_breakpoints()
+                self.set_spyder_breakpoints(breakpoints)
+        except (CommError, TimeoutError):
+            logger.debug("Could not send preloop to frontend.")
+
+
+    def postloop(self):
+        """Notifies spyder that the loop has ended."""
+        try:
+            _frontend_request(blocking=False).set_debug(False)
+        except (CommError, TimeoutError):
+            logger.debug("Could not send postloop to frontend.")
```

Comment:
This is confusing because `postloop` is set in the kernel (and you're doing it below this line).

Are you really informing the frontend that the debug session finished instead? If that's the case, why is that now necessary?

Reply:
I want to know if the kernel is debugging or not, so I know if I need to use pdb to fetch the new messages. I use `u"!get_ipython().kernel.frontend_comm.remote_call(blocking=True).pong()"`, which only works while debugging.

Reply:
Aldo this could be useful in spyder to make decisions on debugging. Now in the spyder code, `_reading` is taken as meaning `_reading_pdb` in many places, but you could have a call to `input()` where it might not be true. It wold therefore be useful to have a way of telling if debugging is really taking place.